make/hotspot/lib/JvmOverrideFiles.gmk

```diff
@@ -37,6 +37,10 @@ ifeq ($(TOOLCHAIN_TYPE), gcc)
     # Need extra inlining to collapse shared marking code into the hot marking loop
     BUILD_LIBJVM_shenandoahMark.cpp_CXXFLAGS := --param inline-unit-growth=1000
   endif
+  # disable lto in g1ParScanThreadState because of special inlining/flattening used there
+  ifeq ($(call check-jvm-feature, link-time-opt), true)
+    BUILD_LIBJVM_g1ParScanThreadState.cpp_CXXFLAGS := -fno-lto
+  endif
```

Comment:
Even with this change I still get ~200 warnings when building with gcc13.2,
mostly -Wattribute-warning. This is consistent with what I reported previously
with ATTRIBUTE_WARNING as a nop. So this change isn't sufficient.

I'm also not sure it's desirable. We don't know the impact of LTO on top of
flattening. It might be beneficial. This change ought to come with some
performance measurements, if made at all. And not under the rubric of a
warnings fix.

Reply:
There are quite a number of things in HotSpot that seemingly wouldn't work with LTO enabled. One notable example is os::current\_stack\_pointer, which on many platforms simply returns the frame pointer of os::current\_stack\_pointer or the address of a local inside os::current\_stack\_pointer. With LTO enabled this becomes eligible for inlining, and the "stack pointer" ends up becoming the frame pointer of the caller, which is not correct. Fixing these issues and making LTO viable is one of the things I hope to do, although not now

Reply:
Hmmm ... I thought we had LTO enabled in the past for Java SE Embedded ...

Reply:
> And not under the rubric of a warnings fix.

Maybe the title is a bit misleading, it is more about a build fix (lto build did not work for me with gcc 11.3.0; with this fix it works).
If the lot stuff is broken and there is no intent to fix it, we might consider removing it (but fixing it would probably be better).

Reply:
I would hope that LTO won't be removed. I plan to try making it viable, and I also use it downstream on Windows. I'm having trouble downgrading my gcc package on WSL to reproduce the warning, so might take me some time to find a better fix

Reply:
> Hmmm ... I thought we had LTO enabled in the past for Java SE Embedded ...

Hi Julian / David,  I  started the HS :tier1 tests with the  lto-enabled JVM (linuxx86_64).
Most tests work, but I get a number of errors too, mostly (but not only) in the  serviceability/sa   area.
A lot of tests fail with such an exception

```
 stderr: [Exception in thread "main" java.lang.InternalError: Metadata does not appear to be polymorphic
	at jdk.hotspot.agent/sun.jvm.hotspot.types.basic.BasicTypeDataBase.findDynamicTypeForAddress(BasicTypeDataBase.java:223)
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.VirtualBaseConstructor.instantiateWrapperFor(VirtualBaseConstructor.java:104)
	at jdk.hotspot.agent/sun.jvm.hotspot.oops.Metadata.instantiateWrapperFor(Metadata.java:78)
	at jdk.hotspot.agent/sun.jvm.hotspot.oops.Oop.getKlassForOopHandle(Oop.java:224)
	at jdk.hotspot.agent/sun.jvm.hotspot.oops.ObjectHeap.newOop(ObjectHeap.java:180)
	at jdk.hotspot.agent/sun.jvm.hotspot.oops.VMOopHandle.resolve(VMOopHandle.java:61)
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.JavaThread.getThreadObj(JavaThread.java:365)
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.JavaThread.getCurrentParkBlocker(JavaThread.java:438)
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.DeadlockDetector.print(DeadlockDetector.java:80)
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.DeadlockDetector.print(DeadlockDetector.java:39)
	at jdk.hotspot.agent/sun.jvm.hotspot.tools.PStack.run(PStack.java:79)
	at jdk.hotspot.agent/sun.jvm.hotspot.CommandProcessor$32.doit(CommandProcessor.java:1229)
	at jdk.hotspot.agent/sun.jvm.hotspot.CommandProcessor.executeCommand(CommandProcessor.java:2230)
	at jdk.hotspot.agent/sun.jvm.hotspot.CommandProcessor.executeCommand(CommandProcessor.java:2200)
	at jdk.hotspot.agent/sun.jvm.hotspot.CommandProcessor.run(CommandProcessor.java:2071)
	at jdk.hotspot.agent/sun.jvm.hotspot.CLHSDB.run(CLHSDB.java:112)
	at jdk.hotspot.agent/sun.jvm.hotspot.CLHSDB.main(CLHSDB.java:44)
	at jdk.hotspot.agent/sun.jvm.hotspot.SALauncher.runCLHSDB(SALauncher.java:285)
	at jdk.hotspot.agent/sun.jvm.hotspot.SALauncher.main(SALauncher.java:507)
```

So I would say that the lto-enabled JVM is not completely broken but has some issues here and there (revealed by the jtreg tests).

Reply:
I switched from my standard gcc 11.3.0 devkit to 13.2.1; build works nicely with it and  jtreg HS :tier1 results look similar.
I still get those ugly  `warning: call to ‘vsnprintf’ declared with attribute warning: use os::vsnprintf`   [-Wattribute-warning] ;  maybe we should disable the attribute warnings for the lto build ?

One thing I maybe have to add - I currently do not build the lto build with gtest (--with-gtest was removed from my script).

Reply:
Yes, I think it's an acceptable workaround to disable attribute\-warning (It's one of the solutions I mentioned earlier, but I'm guessing you mean more aggressively). LTO is not the typical way of compiling HotSpot, so this shouldn't be a problem, until LTO becomes viable enough that it becomes the standard, but that's probably going to be a long time in the future

Reply:
As mentioned in the JBS issue, some of my problems are a result of the devkit I'm using not putting
the lto-plugin in the necessary spot.  So I think I'm now where @MBaesken is at.

I worry that LTO could become the default fairly quickly if it really does show good results, once the
kinds of issues @TheShermanTanker mentioned are fixed.  That would make the poisoning of various
functions effectively cease to work if we disable that feature in LTO builds.

Also, does disabling of LTO for that one file actually work?  The gcc docs say that mixing LTO and
non-LTO objects isn't possible with GNU binutils, and it just falls back to nolto-rel:
https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Link-Options.html#:~:text=During%20the%20incremental,link%2Dtime%20optimization
Maybe it works with the gold linker?  Which seems to be the default in the devkit I'm using.

Reply:
I misspoke a bit earlier, I do not think LTO will become the Standard way of compiling HotSpot, even if all the issues with it are fixed. It simply takes too long, which I think is the main reason Erik mentioned why it is not used for the product today. I intend to make LTO viable, for those who want the extra performance at the expense of link times, but it is unlikely to become the default

The documentation linked above is for incremental linking, it is indeed possible to link LTO and non LTO object files together if linking normally and not incrementally, otherwise LTO would already fail horribly on assembly files (Which are always assembled directly into non LTO object files), and HotSpot contains many assembly source files (This might be an issue for Magnus and his static Java, though, since I think that relies on incremental linking). However, that said, I am not fond of disabling LTO for any specific file, since that will cause a decline in the resulting assembly quality

I've been looking at alternative ways to implement poisoning of methods we don't want people to be using that don't rely on the warning attribute, however that hasn't been successful ([8313396](https://bugs.openjdk.org/browse/JDK-8313396)). I've checked the gcc bug tracker, and the current warnings are apparently a hard to fix bug in gcc, since pragmas are not saved into the GIMPLE when LTO is enabled. I'll list the poisoning thing as a blocker for making LTO viable, so the issues with LTO won't be fixed until the poisoning works with LTO. I'll look into poisoning again once my Windows fixes make it in

As a workaround that doesn't rely on disabling poisoning whatsoever, you could NOINLINE code that contains an ALLOW\_C\_FUNCTION, but that would likewise also inhibit optimizations for the JVM

Reply:
LTO seems to be less than a factor of 2 longer build times. If it makes a
measurable difference in performance, that seems like a clear win for
production builds. Certainly hotspot developers wouldn't want it on, but we
will know about that, just like using ccache and not using precompiled headers
for incremental development.

Thanks for clarifying the incremental linking thing.

The relevant gcc bug regarding pragmas not carrying through to LTO is
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80922
#pragma diagnostic ignored not honoured with -flto
It's been around for a few years now, with no obvious progress.

I failed to understand the proposal for 8313396.  I took another look today,
and I'm still not getting it.  Maybe we should talk about it over there.

Reply:
For the record, I'm seeing 3 warnings, 2 of which are for the same code. Both
are -Walloc-size-larger-than= warnings for calls to operator new[].

In gtestMain.cpp, in init_jvm, `new JavaVMOption[num_jvm_options]` is being
warned about.  The warning is obviously wrong, as the value it is complaining
about uint64_max, while the variable has int type.

The other is in (non-Windows) CreateArgvFromArgs in gtest-death-test.cc, and
is being too conservative about value bounds, so also wrong.

So both of these appear to be LTO bugs.

Of course, it's not possible to suppress these warnings via pragmas, because
of the bug that LTO doesn't honor such.

Reply:
Catching up with the discussion...

LTO has been working before, and was indeed enabled for the old embedded platform. Since then, it has not been regularly tested and has certainly bit-rottet.

I think it is a good thing to try to keep it alive. There is potential for big performance improvements, even if that will only be utilized for special builds. The additional build time is the main culprit here. But it is possible that, as Kim says, the *entire* build is perhaps "just" slowed down by a factor of 2x, and that might be acceptable, even if the Hotspot build is slowed down several magnitudes.

Also, there are many promising efforts going on right now, that is trying to get some or all of the LTO benefits in different ways. In particular, iirc, the LLVM linker lld has some kind of "partial LTO" which, iiuc, tries to pick the most important aspeect of LTOs but implemented in a way that has minimal build performance impact.

It is not entirely clear what correspondence there is between keeping the traditional gcc LTO alive, and being able to use this new clang partial LTO, but as a reasonable first guess it will be helpful to keep it working.

Reply:
The old LTO used to set \-O3 directly on both LDFLAGS and CFLAGS, bypassing OPTIMIZATION variable entirely. Just wanted to put this out there, since the topic of traditional gcc LTO came up. I changed that to set OPTIMIZATION to HIGHEST\_JVM some time ago, and I'm not sure whether my change played any part is breaking LTO

LTO is currently only supported for VC and gcc. I can look into how to implement it for clang, though not now. First I'm committed to fixing all the issues with LTO, and making sure Kim's worry of poisoning virtually disappearing when LTO is enabled doesn't become a reality

Reply:
Agree, we should support LTO for gcc AND clang on Linux/macOS.
Regarding the longer link/build times , I can add some info how much longer it takes in our environment (SUSE Linux/gcc).

Reply:
I can add some info here: On Windows, with LTO enabled it takes gcc about 15\-20 minutes to link the JVM, without it takes anywhere between 8\-10 minutes to do so. ld is very slow on Windows due to a bug, so the numbers are skewed towards the high end. I don't have any numbers for VC, but I can assume that even LTO with VC and link.exe is very fast, because VC in general is an extremely fast compiler